File: coffeeshop/src/main/java/com/coffeeshop/coffeeshop/auth/KeycloakAdminClient.java
package com.coffeeshop.coffeeshop.auth;

import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Component;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;
import org.springframework.web.client.HttpStatusCodeException;
import org.springframework.web.client.RestClient;

import java.net.URI;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@Component
public class KeycloakAdminClient {

    private final KeycloakProperties properties;
    private final RestClient restClient = RestClient.create();

    public KeycloakAdminClient(final KeycloakProperties properties) {
        this.properties = properties;
    }

    /**
     * Deletes a realm user; ignores failures so callers can compensate without masking the original error.
     */
    public void deleteUserBestEffort(final UUID userId) {
        try {
            deleteUserBestEffort(userId, obtainAdminAccessToken());
        } catch (final KeycloakAuthException ignored) {
        }
    }

    private void deleteUserBestEffort(final UUID userId, final String adminBearerToken) {
        try {
            restClient.delete()
                    .uri(properties.adminUsersBaseUri() + "/" + userId)
                    .header(HttpHeaders.AUTHORIZATION, "Bearer " + adminBearerToken)
                    .retrieve()
                    .toBodilessEntity();
        } catch (final HttpStatusCodeException ignored) {
        }
    }

    public UUID createUserWithRealmRole(
            final String email,
            final String password,
            final String name,
            final String realmRoleName) {
        final String adminToken = obtainAdminAccessToken();
        final Map<String, Object> payload = new HashMap<>();
        payload.put("username", email);
        payload.put("email", email);
        payload.put("firstName", name);
        payload.put("lastName", name);
        payload.put("enabled", true);
        payload.put("emailVerified", true);
        payload.put("credentials", List.of(Map.of(
                "type", "password",
                "value", password,
                "temporary", false)));

        final UUID userId;
        try {
            final var response = restClient.post()
                    .uri(properties.adminUsersBaseUri())
                    .header(HttpHeaders.AUTHORIZATION, "Bearer " + adminToken)
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(payload)
                    .retrieve()
                    .toBodilessEntity();
            final URI location = response.getHeaders().getLocation();
            if (location == null) {
                throw new KeycloakAuthException("Keycloak user create response missing Location");
            }
            final String path = location.getPath();
            final String idSegment = path.substring(path.lastIndexOf('/') + 1);
            userId = UUID.fromString(idSegment);
        } catch (final HttpStatusCodeException ex) {
            if (ex.getStatusCode().value() == 409) {
                throw new KeycloakAuthException("User already exists in Keycloak", ex);
            }
            throw new KeycloakAuthException("Keycloak admin request failed: " + ex.getStatusCode(), ex);
        }

        try {
            final Map<String, Object> roleRep = restClient.get()
                    .uri(properties.adminRoleUri(realmRoleName))
                    .header(HttpHeaders.AUTHORIZATION, "Bearer " + adminToken)
                    .retrieve()
                    .body(new ParameterizedTypeReference<Map<String, Object>>() {
                    });
            if (roleRep == null) {
                throw new KeycloakAuthException("Realm role not found: " + realmRoleName);
            }
            restClient.post()
                    .uri(properties.adminUsersBaseUri() + "/" + userId + "/role-mappings/realm")
                    .header(HttpHeaders.AUTHORIZATION, "Bearer " + adminToken)
                    .contentType(MediaType.APPLICATION_JSON)
                    .body(List.of(roleRep))
                    .retrieve()
                    .toBodilessEntity();

            return userId;
        } catch (final HttpStatusCodeException ex) {
            deleteUserBestEffort(userId, adminToken);
            throw new KeycloakAuthException("Keycloak admin request failed: " + ex.getStatusCode(), ex);
        } catch (final KeycloakAuthException ex) {
            deleteUserBestEffort(userId, adminToken);
            throw ex;
        }
    }

    private String obtainAdminAccessToken() {
        final MultiValueMap<String, String> form = new LinkedMultiValueMap<>();
        form.add("grant_type", "password");
        form.add("client_id", "admin-cli");
        form.add("username", properties.getAdminUsername());
        form.add("password", properties.getAdminPassword());
        try {
            final Map<?, ?> body = restClient.post()
                    .uri(properties.adminTokenUri())
                    .contentType(MediaType.APPLICATION_FORM_URLENCODED)
                    .body(form)
                    .retrieve()
                    .body(Map.class);
            if (body == null || body.get("access_token") == null) {
                throw new KeycloakAuthException("Keycloak admin token missing access_token");
            }
            return body.get("access_token").toString();
        } catch (final HttpStatusCodeException ex) {
            throw new KeycloakAuthException("Keycloak admin login failed: " + ex.getStatusCode(), ex);
        }
    }
}

File: coffeeshop-go/internal/auth/keycloak_admin_client.go
package auth

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/mastilovic/coffeeshop-go/internal/config"
)

type KeycloakAdminClient struct {
	cfg        config.Config
	httpClient *http.Client
}

func NewKeycloakAdminClient(cfg config.Config) *KeycloakAdminClient {
	return &KeycloakAdminClient{
		cfg:        cfg,
		httpClient: &http.Client{Timeout: 10 * time.Second},
	}
}

func (c *KeycloakAdminClient) CreateUserWithRealmRole(email, password, name, realmRoleName string) (string, error) {
	adminToken, err := c.obtainAdminAccessToken()
	if err != nil {
		return "", err
	}

	payload := map[string]interface{}{
		"username":      email,
		"email":         email,
		"firstName":     name,
		"lastName":      name,
		"enabled":       true,
		"emailVerified": true,
		"credentials": []map[string]interface{}{
			{"type": "password", "value": password, "temporary": false},
		},
	}

	body, _ := json.Marshal(payload)
	usersURI := c.adminUsersBaseURI()
	req, _ := http.NewRequest(http.MethodPost, usersURI, bytes.NewReader(body))
	req.Header.Set("Authorization", "Bearer "+adminToken)
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("keycloak create user request failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == 409 {
		return "", &KeycloakAuthError{StatusCode: 409}
	}
	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("keycloak create user failed with status %d", resp.StatusCode)
	}

	location := resp.Header.Get("Location")
	if location == "" {
		return "", fmt.Errorf("keycloak user create response missing Location header")
	}
	parts := strings.Split(location, "/")
	userID := parts[len(parts)-1]

	if err := c.assignRealmRole(userID, realmRoleName, adminToken); err != nil {
		c.deleteUserBestEffort(userID, adminToken)
		return "", err
	}

	return userID, nil
}

func (c *KeycloakAdminClient) DeleteUserBestEffort(userID string) {
	adminToken, err := c.obtainAdminAccessToken()
	if err != nil {
		return
	}
	c.deleteUserBestEffort(userID, adminToken)
}

func (c *KeycloakAdminClient) deleteUserBestEffort(userID, adminToken string) {
	req, _ := http.NewRequest(http.MethodDelete, c.adminUsersBaseURI()+"/"+userID, nil)
	req.Header.Set("Authorization", "Bearer "+adminToken)
	resp, err := c.httpClient.Do(req)
	if err == nil {
		resp.Body.Close()
	}
}

func (c *KeycloakAdminClient) assignRealmRole(userID, roleName, adminToken string) error {
	roleURI := c.cfg.KeycloakBaseURL + "/admin/realms/" + c.cfg.KeycloakRealm + "/roles/" + roleName
	req, _ := http.NewRequest(http.MethodGet, roleURI, nil)
	req.Header.Set("Authorization", "Bearer "+adminToken)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("failed to fetch role: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return fmt.Errorf("realm role not found: %s", roleName)
	}

	var roleRep map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&roleRep); err != nil {
		return fmt.Errorf("failed to decode role response: %w", err)
	}

	roleBody, _ := json.Marshal([]interface{}{roleRep})
	mappingURI := c.adminUsersBaseURI() + "/" + userID + "/role-mappings/realm"
	req2, _ := http.NewRequest(http.MethodPost, mappingURI, bytes.NewReader(roleBody))
	req2.Header.Set("Authorization", "Bearer "+adminToken)
	req2.Header.Set("Content-Type", "application/json")

	resp2, err := c.httpClient.Do(req2)
	if err != nil {
		return fmt.Errorf("failed to assign role: %w", err)
	}
	defer resp2.Body.Close()

	if resp2.StatusCode >= 400 {
		return fmt.Errorf("keycloak role assignment failed with status %d", resp2.StatusCode)
	}
	return nil
}

func (c *KeycloakAdminClient) obtainAdminAccessToken() (string, error) {
	adminTokenURI := c.cfg.KeycloakBaseURL + "/realms/master/protocol/openid-connect/token"
	form := url.Values{
		"grant_type": {"password"},
		"client_id":  {"admin-cli"},
		"username":   {c.cfg.KeycloakAdminUser},
		"password":   {c.cfg.KeycloakAdminPassword},
	}

	resp, err := c.httpClient.Post(adminTokenURI, "application/x-www-form-urlencoded", strings.NewReader(form.Encode()))
	if err != nil {
		return "", fmt.Errorf("keycloak admin login failed: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return "", fmt.Errorf("keycloak admin login failed with status %d", resp.StatusCode)
	}

	var body map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		return "", fmt.Errorf("failed to decode admin token response: %w", err)
	}

	token, _ := body["access_token"].(string)
	if token == "" {
		return "", fmt.Errorf("keycloak admin token missing access_token")
	}
	return token, nil
}

func (c *KeycloakAdminClient) adminUsersBaseURI() string {
	return c.cfg.KeycloakBaseURL + "/admin/realms/" + c.cfg.KeycloakRealm + "/users"
}
